### DISCUSSION TITLE
Support for Gamma Ramps(?)

### DISCUSSION THREAD

**Discussion by stellarpower** [any1/wayvnc#313] [Q&A]
Hi,

Wayvnc just works - thanks!

One of the main reasons I'm using a wlroots compositor at the moment is support for performing gamma ramps using the [zwlr_gamma_... protocol extensions](https://wayland.app/protocols/wlr-gamma-control-unstable-v1). See [gammastep]( https://gitlab.com/chinstrap/gammastep ) for an example.

As I expected, when I connect via VNC, I can see the screen fine, but the gamma isn't applied to the virtual copy of the output - I see a copy of my hardware screen on my remote computer, but, the scree brightness us unaffected whilst the hardware output does have the ramp applied. The same has happened when taking screenshots and sharing through browser/screen recording before.

Do you know if wayvnc would be able to support this at all? Or will it necessarily be limited to the physical output? I'm not massively clued up on how wayland works under the bonnet, but I've taken a brief look at the implementation within wayfire, so wouldn't be enormously surprised if the way wayvnc hooks into the stack would mean it is unable to see the rendered frames.

Thanks

**Comment by any1** [comment]
The gamma settings are not baked into frame buffers; they're passed on to the GPU. I think that the GPU might even pass them on to the monitor (but I'm not sure about that bit), in which case, not even DRM write-back will give you an image with the gamma applied.

**Reply by stellarpower** [reply]
Oh really? That's interesting, thanks. I didn't realise it was performing gamma in the same way as say managing a colour profile. Do you think DRM write-back could be used to achieve a similar effect? In my case I'm largely concerned about brightness - simply multiplying each (r,g,b) value by a constant once per frame (or adding an offset) would work fine. I assumed that I'd need to patch the compositor to achieve that for a full screen, and for security reasons it would be "tricky" just to get a pointer to the full buffer for the current frame (even if root), but I don't know if it's an exposed protocol.

Although this doesn't help with the VNC aspect, but if it's going across the network I guess that means wayvnc does have access to its full buffer somewhere and I could simply patch it to modify the values as needed before they go on the wire and do away with the Gamma protocol. (My screen's knackered!)

**Reply by any1** [reply]
I only have surface-level knowledge about this, so I can't tell you for sure if DRM write-back will make any difference unless I study it.

You can change the code that samples pixel buffers so that it modifies the color, but I think that a better way to do it would be to modify the shaders in the compositor.
